File: app/src/main/java/il/ronmad/speedruntimer/TimerService.java
package il.ronmad.speedruntimer;

import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.app.Service;
import android.content.Intent;
import android.content.SharedPreferences;
import android.graphics.PixelFormat;
import android.graphics.Typeface;
import android.os.Build;
import android.os.IBinder;
import android.preference.PreferenceManager;
import android.support.v4.app.NotificationCompat;
import android.support.v4.content.ContextCompat;
import android.support.v7.app.AlertDialog;
import android.util.DisplayMetrics;
import android.view.Gravity;
import android.view.MotionEvent;
import android.view.View;
import android.view.Window;
import android.view.WindowManager;
import android.widget.TextView;

import io.realm.Realm;
import io.realm.RealmChangeListener;

public class TimerService extends Service {

    public static boolean IS_ACTIVE = false;

    private Realm realm;
    private RealmChangeListener<Realm> realmChangeListener;
    private SharedPreferences prefs;
    private String gameName;
    private String categoryName;
    private Game game;
    private Category category;

    private View mView;
    private Chronometer chronometer;

    private NotificationManager notificationManager;
    private NotificationCompat.Builder notificationBuilder;

    private WindowManager mWindowManager;
    private boolean moved;

    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }

    @Override
    public void onCreate() {
        super.onCreate();
        realm = Realm.getDefaultInstance();
        realmChangeListener = realm -> onDataChange();
        realm.addChangeListener(realmChangeListener);
    }

    @SuppressWarnings("ConstantConditions")
    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        IS_ACTIVE = true;

        gameName = intent.getStringExtra(getString(R.string.extra_game));
        categoryName = intent.getStringExtra(getString(R.string.extra_category));
        game = realm.where(Game.class).equalTo("name", gameName).findFirst();
        category = game.getCategory(categoryName);

        Notification notification = setupNotification();
        startForeground(R.integer.notification_id, notification);

        prefs = PreferenceManager.getDefaultSharedPreferences(this);
        Chronometer.bestTime = category.bestTime;
        Chronometer.colorNeutral = prefs.getInt(getString(R.string.key_pref_color_neutral),
                ContextCompat.getColor(this, R.color.colorTimerNeutralDefault));
        Chronometer.colorAhead = prefs.getInt(getString(R.string.key_pref_color_ahead),
                ContextCompat.getColor(this, R.color.colorTimerAheadDefault));
        Chronometer.colorBehind = prefs.getInt(getString(R.string.key_pref_color_behind),
                ContextCompat.getColor(this, R.color.colorTimerBehindDefault));
        Chronometer.colorPB = prefs.getInt(getString(R.string.key_pref_color_pb),
                ContextCompat.getColor(this, R.color.colorTimerPBDefault));
        Chronometer.countdown = prefs.getLong(getString(R.string.key_pref_timer_countdown), 0L);
        Chronometer.showMillis = prefs.getBoolean(getString(R.string.key_pref_timer_show_millis), true);

        mWindowManager = (WindowManager) getSystemService(WINDOW_SERVICE);
        setupLayoutComponents();
        setupView();

        return super.onStartCommand(intent, flags, startId);
    }

    @Override
    public void onDestroy() {
        realm.removeChangeListener(realmChangeListener);
        if (mView != null) {
            if (game != null) {
                game.timerPosition.set(mWindowParams.x, mWindowParams.y);
            }
            mWindowManager.removeView(mView);
        }
        realm.close();
        IS_ACTIVE = false;
        super.onDestroy();
    }

    private void onDataChange() {
        Chronometer.bestTime = category.bestTime;
        notificationBuilder.setContentText(category.bestTime > 0 ?
                String.format("PB: %s", Util.getFormattedTime(category.bestTime))
                : null);
        notificationManager.notify(R.integer.notification_id, notificationBuilder.build());
    }

    private Notification setupNotification() {
        notificationManager = (NotificationManager) getSystemService(NOTIFICATION_SERVICE);
        notificationBuilder = new NotificationCompat.Builder(this, getString(R.string.notification_channel_id))
                .setSmallIcon(Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP ?
                        R.drawable.ic_timer_black_48dp : R.drawable.ic_stat_timer)
                .setContentTitle(String.format("%s %s", game.name, category.name))
                .setContentIntent(PendingIntent.getActivity(this, 0,
                        new Intent(this, MainActivity.class),
                        PendingIntent.FLAG_UPDATE_CURRENT))
                .addAction(Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP ?
                                R.drawable.ic_close_black_24dp : R.drawable.ic_stat_close,
                        getString(R.string.close_timer),
                        PendingIntent.getBroadcast(this, 0,
                                new Intent(getString(R.string.action_close_timer)),
                                PendingIntent.FLAG_UPDATE_CURRENT))
                .setAutoCancel(true)
                .setOnlyAlertOnce(true);
        if (category.bestTime > 0) {
            notificationBuilder.setContentText(String.format("PB: %s", Util.getFormattedTime(category.bestTime)));
        }
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            NotificationChannel notificationChannel = new NotificationChannel(
                    getString(R.string.notification_channel_id),
                    getString(R.string.app_name),
                    NotificationManager.IMPORTANCE_LOW);
            notificationChannel.enableVibration(false);
            notificationChannel.enableLights(false);
            notificationManager.createNotificationChannel(notificationChannel);
        }
        Notification notification = notificationBuilder.build();
        notificationManager.notify(R.integer.notification_id, notification);
        return notification;
    }

    private void setupLayoutComponents() {
        setTheme(R.style.AppTheme);
        mView = View.inflate(this, R.layout.timer_overlay, null);

        mView.setBackgroundColor(prefs.getInt(getString(R.string.key_pref_color_background),
                ContextCompat.getColor(this, R.color.colorTimerBackgroundDefault)));
        int size = Integer.parseInt(prefs.getString(getString(R.string.key_pref_timer_size), "32"));
        TextView chronoRest = mView.findViewById(R.id.chronoRest);
        TextView chronoMillis = mView.findViewById(R.id.chronoMillis);
        chronoRest.setTextSize(size);
        chronoMillis.setTextSize((float) (size * 0.75));

        chronoRest.setTypeface(Typeface.createFromAsset(
                getAssets(), "fonts/digital-7.ttf"));
        chronoMillis.setTypeface(Typeface.createFromAsset(
                getAssets(), "fonts/digital-7.ttf"));

        chronometer = new Chronometer(this, mView);

        mView.setLongClickable(true);
        mView.setOnTouchListener(new View.OnTouchListener() {
            private DisplayMetrics metrics = new DisplayMetrics();
            private int initialX;
            private int initialY;
            private float initialTouchX;
            private float initialTouchY;

            long touchTime;
            long startTime = System.currentTimeMillis();

            @Override
            public boolean onTouch(View v, MotionEvent event) {
                if (System.currentTimeMillis() - startTime <= 300) {
                    return false;
                }

                mWindowManager.getDefaultDisplay().getMetrics(metrics);

                switch (event.getAction()) {
                    case MotionEvent.ACTION_DOWN:
                        initialX = mWindowParams.x;
                        initialY = mWindowParams.y;
                        initialTouchX = event.getRawX();
                        initialTouchY = event.getRawY();
                        touchTime = System.currentTimeMillis();
                        moved = false;
                        break;
                    case MotionEvent.ACTION_UP:
                        if (moved) break;
                        if (System.currentTimeMillis() - touchTime < 250) {
                            if (chronometer.isRunning()) {
                                chronometer.stop();
                            } else {
                                chronometer.start();
                            }
                        }
                        break;
                    case MotionEvent.ACTION_MOVE:
                        int targetX = initialX - (int) (event.getRawX() - initialTouchX);
                        int targetY = initialY - (int) (event.getRawY() - initialTouchY);
                        targetX = Math.max(0, Math.min(targetX, metrics.widthPixels - v.getWidth()));
                        targetY = Math.max(0, Math.min(targetY, metrics.heightPixels - v.getHeight()));
                        if (!moved &&
                                Math.pow(targetX - initialX, 2) + Math.pow(targetY - initialY, 2) < 25*25)
                            break;
                        moved = true;
                        mWindowParams.x = targetX;
                        mWindowParams.y = targetY;
                        mWindowManager.updateViewLayout(mView, mWindowParams);
                        break;
                }
                v.performClick();
                return false;
            }
        });

        mView.setOnLongClickListener(view -> {
            long time = chronometer.getTimeElapsed();
            if (moved || !Chronometer.started || Chronometer.running || time == 0) {
                return false;
            }
            if (time < 0) {
                chronometer.reset();
            } else if (category.bestTime > 0 && time >= category.bestTime) {
                chronometer.reset();
                category.incrementRunCount();
            } else if (!Chronometer.running) {
                AlertDialog resetDialog = new AlertDialog.Builder(TimerService.this)
                        .setTitle(category.bestTime == 0 ? "New personal best!" :
                                String.format("New personal best! (%s)",
                                        Util.getFormattedTime(time - category.bestTime)))
                        .setMessage("Save it?")
                        .setPositiveButton(R.string.save_reset, (dialogInterface, i) -> {
                            chronometer.reset();
                            category.setData(time, category.runCount + 1);
                        })
                        .setNegativeButton(R.string.reset, (dialogInterface, i) ->  {
                            chronometer.reset();
                            category.incrementRunCount();
                        })
                        .setNeutralButton(android.R.string.cancel, null)
                        .create();
                Window window = resetDialog.getWindow();
                if (window != null) {
                    window.setType(Build.VERSION.SDK_INT >= Build.VERSION_CODES.O ?
                            WindowManager.LayoutParams.TYPE_APPLICATION_OVERLAY
                            : WindowManager.LayoutParams.TYPE_SYSTEM_ALERT);
                }
                resetDialog.show();
            }
            return true;
        });
    }

    WindowManager.LayoutParams mWindowParams;
    private void setupView() {
        DisplayMetrics metrics = new DisplayMetrics();
        mWindowManager.getDefaultDisplay().getMetrics(metrics);
        mWindowParams = new WindowManager.LayoutParams(
                WindowManager.LayoutParams.WRAP_CONTENT,
                WindowManager.LayoutParams.WRAP_CONTENT,
                Build.VERSION.SDK_INT >= Build.VERSION_CODES.O ? WindowManager.LayoutParams.TYPE_APPLICATION_OVERLAY
                : WindowManager.LayoutParams.TYPE_PHONE,
                WindowManager.LayoutParams.FLAG_NOT_TOUCH_MODAL | WindowManager.LayoutParams.FLAG_WATCH_OUTSIDE_TOUCH |
                        WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE,
                PixelFormat.TRANSLUCENT);
        mWindowParams.gravity = Gravity.BOTTOM | Gravity.END;

        int x = game.getTimerPosition().x;
        int y = game.getTimerPosition().y;
        mWindowParams.x = Math.max(0, Math.min(x, metrics.widthPixels - mWindowParams.width));
        mWindowParams.y = Math.max(0, Math.min(y, metrics.heightPixels - mWindowParams.height));
        mWindowManager.addView(mView, mWindowParams);
    }
}

File: app/src/main/kotlin/il/ronmad/speedruntimer/TimerService.kt
package il.ronmad.speedruntimer

import android.annotation.SuppressLint
import android.annotation.TargetApi
import android.app.*
import android.content.*
import android.graphics.PixelFormat
import android.graphics.Typeface
import android.os.Build
import android.os.IBinder
import android.util.DisplayMetrics
import android.view.*
import androidx.core.app.NotificationCompat
import androidx.preference.PreferenceManager
import com.afollestad.materialdialogs.MaterialDialog
import com.afollestad.materialdialogs.checkbox.checkBoxPrompt
import com.afollestad.materialdialogs.checkbox.isCheckPromptChecked
import il.ronmad.speedruntimer.activities.MainActivity
import il.ronmad.speedruntimer.databinding.TimerOverlayBinding
import il.ronmad.speedruntimer.realm.*
import io.realm.Realm
import io.realm.RealmChangeListener
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlin.math.pow

class TimerService : Service() {

    private lateinit var realm: Realm
    private lateinit var realmChangeListener: RealmChangeListener<Realm>
    lateinit var prefs: SharedPreferences
    private lateinit var chronometer: Chronometer
    private lateinit var category: Category
    private var splitsIter: MutableListIterator<Split>? = null
    private val segmentTimes: MutableList<Long> = mutableListOf()
    private var currentSplitStartTime = 0L
    private var hasSplits = false
    private var comparison: Comparison = Comparison.PERSONAL_BEST

    private lateinit var notificationManager: NotificationManager
    private lateinit var notificationBuilder: NotificationCompat.Builder
    private lateinit var receiver: BroadcastReceiver

    lateinit var mBinding: TimerOverlayBinding
    private lateinit var mWindowManager: WindowManager
    private lateinit var mWindowParams: WindowManager.LayoutParams
    private var moved = false

    private var startedProperly = false

    override fun onBind(intent: Intent?): IBinder? = null

    override fun onCreate() {
        super.onCreate()
        realm = Realm.getDefaultInstance()
        realmChangeListener = RealmChangeListener { onDataChange() }
        realm.addChangeListener(realmChangeListener)
    }

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        IS_ACTIVE = true

        intent?.let {
            gameName = it.getStringExtra(getString(R.string.extra_game)).orEmpty()
            categoryName = it.getStringExtra(getString(R.string.extra_category)).orEmpty()
        }
        if (gameName.isEmpty() || categoryName.isEmpty()) {
            stopSelf()
            startedProperly = false
            return START_NOT_STICKY
        }

        category = realm.getCategoryByName(gameName, categoryName)!!
        hasSplits = category.splits.isNotEmpty()

        setupReceiver()
        val notification = setupNotification()
        startForeground(R.integer.notification_id, notification)

        setupChronometerPrefs()

        mWindowManager = getSystemService(Context.WINDOW_SERVICE) as WindowManager
        setupLayoutComponents()
        setupView()

        startedProperly = true
        IS_ACTIVE = true
        return super.onStartCommand(intent, flags, startId)
    }

    override fun onDestroy() {
        realm.removeChangeListener(realmChangeListener)
        if (startedProperly) {
            mWindowManager.removeView(mBinding.root)
            unregisterReceiver(receiver)
        }
        realm.close()
        IS_ACTIVE = false
        super.onDestroy()
    }

    fun closeTimer(fromOnResume: Boolean) {
        if (Chronometer.started) {
            Dialogs.showTimerActiveDialog(this, fromOnResume) { stopSelf() }
        } else {
            stopSelf()
        }
    }

    private fun onDataChange() {
        notificationBuilder.setContentText(
            if (category.bestTime > 0)
                "PB: ${category.bestTime.getFormattedTime()}"
            else
                null
        )
        notificationManager.notify(R.integer.notification_id, notificationBuilder.build())
    }

    private fun setupChronometerPrefs() {
        prefs = PreferenceManager.getDefaultSharedPreferences(this)
        Chronometer.apply {
            colorNeutral = prefs.getInt(
                getString(R.string.key_pref_color_neutral),
                getColorCpt(R.color.colorTimerNeutralDefault)
            )
            colorAhead = prefs.getInt(
                getString(R.string.key_pref_color_ahead),
                getColorCpt(R.color.colorTimerAheadDefault)
            )
            colorBehind = prefs.getInt(
                getString(R.string.key_pref_color_behind),
                getColorCpt(R.color.colorTimerBehindDefault)
            )
            colorPB = prefs.getInt(
                getString(R.string.key_pref_color_pb),
                getColorCpt(R.color.colorTimerPBDefault)
            )
            colorBestSegment = prefs.getInt(
                getString(R.string.key_pref_color_best_segment),
                getColorCpt(R.color.colorTimerBestSegmentDefault)
            )
            countdown = prefs.getLong(getString(R.string.key_pref_timer_countdown), 0L)
            showMillis = prefs.getBoolean(getString(R.string.key_pref_timer_show_millis), true)
            alwaysMinutes = prefs.getBoolean(getString(R.string.key_pref_timer_always_minutes), true)
        }
        comparison = getComparison()
    }

    private fun setupReceiver() {
        receiver = object : BroadcastReceiver() {
            override fun onReceive(context: Context, intent: Intent) {
                when (intent.action) {
                    getString(R.string.action_close_timer) ->
                        closeTimer(intent.getBooleanExtra(getString(R.string.extra_close_timer_from_onresume), true))
                }
            }
        }
        val intentFilter = IntentFilter()
        intentFilter.addAction(getString(R.string.action_close_timer))
        registerReceiver(receiver, intentFilter)
    }

    private fun setupNotification(): Notification {
        notificationManager = getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
        notificationBuilder = NotificationCompat.Builder(this, getString(R.string.notification_channel_id))
            .setSmallIcon(R.drawable.ic_timer_black_48dp)
            .setContentTitle("${category.gameName} ${category.name}")
            .setContentIntent(
                PendingIntent.getActivity(
                    this, 0,
                    Intent(this, MainActivity::class.java),
                    PendingIntent.FLAG_UPDATE_CURRENT
                )
            )
            .addAction(
                R.drawable.ic_close_black_24dp,
                getString(R.string.close_timer),
                PendingIntent.getBroadcast(
                    this, 0,
                    Intent(getString(R.string.action_close_timer)).also {
                        it.putExtra(getString(R.string.extra_close_timer_from_onresume), false)
                    },
                    PendingIntent.FLAG_UPDATE_CURRENT
                )
            )
            .setAutoCancel(true)
            .setOnlyAlertOnce(true)
        if (category.bestTime > 0) {
            notificationBuilder.setContentText("PB: ${category.bestTime.getFormattedTime()}")
        }
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            setupNotificationChannel()
        }
        return notificationBuilder.build().also { notification ->
            notificationManager.notify(R.integer.notification_id, notification)
        }
    }

    @TargetApi(Build.VERSION_CODES.O)
    private fun setupNotificationChannel() {
        val notificationChannel = NotificationChannel(
            getString(R.string.notification_channel_id),
            getString(R.string.app_name),
            NotificationManager.IMPORTANCE_LOW
        ).also {
            it.enableVibration(false)
            it.enableLights(false)
        }
        notificationManager.createNotificationChannel(notificationChannel)
    }

    @SuppressLint("ClickableViewAccessibility")
    private fun setupLayoutComponents() {
        setTheme(R.style.AppTheme)
        mBinding = TimerOverlayBinding.inflate(LayoutInflater.from(this))

        setupDisplayPrefs()

        chronometer = Chronometer(mBinding)

        mBinding.root.isLongClickable = true
        mBinding.root.setOnTouchListener(object : View.OnTouchListener {
            private val metrics = DisplayMetrics()
            private var initialX: Int = 0
            private var initialY: Int = 0
            private var initialTouchX: Float = 0.toFloat()
            private var initialTouchY: Float = 0.toFloat()

            var touchTime: Long = 0
            var startTime = System.currentTimeMillis()
            var prevTapTime: Long = 0

            override fun onTouch(v: View, event: MotionEvent): Boolean {
                if (System.currentTimeMillis() - startTime <= 300) {
                    return false
                }

                mWindowManager.defaultDisplay.getMetrics(metrics)

                when (event.action) {
                    MotionEvent.ACTION_DOWN -> {
                        initialX = mWindowParams.x
                        initialY = mWindowParams.y
                        initialTouchX = event.rawX
                        initialTouchY = event.rawY
                        touchTime = System.currentTimeMillis()
                        moved = false
                    }
                    MotionEvent.ACTION_UP -> {
                        if (moved)
                            category.getGame().getPosition().set(mWindowParams.x, mWindowParams.y)
                        if (moved || System.currentTimeMillis() - touchTime >= 250) return false
                        // Prevent accidental double-clicking
                        val currentTime = System.currentTimeMillis()
                        if (currentTime - prevTapTime < 400) return false
                        prevTapTime = currentTime

                        val splitTime = chronometer.timeElapsed
                        if (Chronometer.running) {
                            if (splitTime < 0) return false
                            if (hasSplits) {
                                // Split, or stop if on final split.
                                val segmentTime = splitTime - currentSplitStartTime
                                segmentTimes += segmentTime
                                if (prefs.getBoolean(getString(R.string.key_pref_timer_show_delta), true))
                                    getCurrentSplit()?.let { updateDelta(it, splitTime) }

                                if (splitsIter?.hasNext() == true) {
                                    timerSplit()
                                } else {
                                    timerStop()
                                }
                            } else {
                                // No splits, so just stop.
                                timerStop()
                            }
                        } else {
                            if (splitsIter == null)
                                timerStart()
                        }
                        currentSplitStartTime = splitTime.coerceAtLeast(0L)
                    }
                    MotionEvent.ACTION_MOVE -> {
                        val targetX = (initialX - (event.rawX - initialTouchX).toInt())
                            .coerceIn(0, metrics.widthPixels - v.width)
                        val targetY = (initialY - (event.rawY - initialTouchY).toInt())
                            .coerceIn(0, metrics.heightPixels - v.height)
                        if (moved || (targetX - initialX).toDouble().pow(2.0) + (targetY - initialY).toDouble()
                                .pow(2.0) >= 30 * 30
                        ) {
                            moved = true
                            mWindowParams.x = targetX
                            mWindowParams.y = targetY
                            mWindowManager.updateViewLayout(mBinding.root, mWindowParams)
                        }
                    }
                }
                v.performClick()
                return false
            }
        })

        mBinding.root.setOnLongClickListener {
            val time = chronometer.timeElapsed
            val updateData = prefs.getBoolean(getString(R.string.key_pref_save_time_data), true)
            when {
                moved -> return@setOnLongClickListener false
                time <= 0 -> timerReset(updateData = false)
                Chronometer.running || (category.bestTime > 0 && category.bestTime in 0..time) ->
                    timerReset(updateData = updateData)
                category.bestTime == 0L -> timerReset(time, updateData = updateData)
                !updateData -> timerReset(updateData = false)
                else -> {
                    MaterialDialog(this).show {
                        title(text = "New personal best!")
                        val format = "%-16s%s"
                        message(
                            text = format.format("Previous PB:", category.bestTime.getFormattedTime()) +
                                    "\n" + format.format("Improvement:", (time - category.bestTime).getFormattedTime())
                        )
                        checkBoxPrompt(text = if (hasSplits) "Save splits" else "Save", isCheckedDefault = true) {}
                        positiveButton(R.string.reset) {
                            timerReset(time, updateData = isCheckPromptChecked())
                        }
                        negativeButton(android.R.string.cancel)
                        window?.setType(
                            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O)
                                WindowManager.LayoutParams.TYPE_APPLICATION_OVERLAY
                            else
                                WindowManager.LayoutParams.TYPE_SYSTEM_ALERT
                        )
                    }
                }
            }
            true
        }
    }

    private fun setupDisplayPrefs() {
        mBinding.root.setBackgroundColor(
            prefs.getInt(
                getString(R.string.key_pref_color_background),
                getColorCpt(R.color.colorTimerBackgroundDefault)
            )
        )

        setupSize()
        setupFont()

        mBinding.currentSplit.setTextColor(Chronometer.colorNeutral)
    }

    private fun setupSize() {
        val timerSizesVals = resources.getStringArray(R.array.timer_sizes_values)
        val size = prefs.getString(getString(R.string.key_pref_timer_size), timerSizesVals[1])!!.toFloat()
        val millisSize = size * 0.75f
        val deltaSize = size * 0.375f
        val splitSize = size * 0.5f
        mBinding.apply {
            chronoViewSet.forEach {
                it.textSize = size
            }
            chronoDot.textSize = millisSize
            chronoMilli2.textSize = millisSize
            chronoMilli1.textSize = millisSize
            delta.textSize = deltaSize
            currentSplit.textSize = splitSize
        }
    }

    private fun setupFont() {
        val digital7Font = Typeface.createFromAsset(assets, "fonts/digital-7.ttf")
        mBinding.chronoViewSet.forEach { it.typeface = digital7Font }
    }

    private fun setupView() {
        val metrics = DisplayMetrics()
        mWindowManager.defaultDisplay.getMetrics(metrics)
        mWindowParams = WindowManager.LayoutParams(
            WindowManager.LayoutParams.WRAP_CONTENT,
            WindowManager.LayoutParams.WRAP_CONTENT,
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O)
                WindowManager.LayoutParams.TYPE_APPLICATION_OVERLAY
            else
                WindowManager.LayoutParams.TYPE_PHONE,
            WindowManager.LayoutParams.FLAG_NOT_TOUCH_MODAL or WindowManager.LayoutParams.FLAG_WATCH_OUTSIDE_TOUCH or
                    WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE,
            PixelFormat.TRANSLUCENT
        )
        mWindowParams.gravity = Gravity.BOTTOM or Gravity.END

        mWindowManager.addView(mBinding.root, mWindowParams)
        resetTimerPosition()
    }

    private fun resetTimerPosition() {
        val (x, y) = category.getGame().getPosition()
        val metrics = DisplayMetrics()
        mWindowManager.defaultDisplay.getMetrics(metrics)
        mWindowParams.x = x.coerceIn(0, metrics.widthPixels - mWindowParams.width)
        mWindowParams.y = y.coerceIn(0, metrics.heightPixels - mWindowParams.height)
        mWindowManager.updateViewLayout(mBinding.root, mWindowParams)
    }

    private fun timerReset(newPB: Long = 0L, updateData: Boolean = true) {
        chronometer.reset()
        if (updateData) {
            if (newPB == 0L) {
                category.incrementRunCount()
                category.updateSplits(segmentTimes, false)
            } else {
                category.updateData(bestTime = newPB, runCount = category.runCount + 1)
                category.updateSplits(segmentTimes, true)
                FSTWidget.forceUpdateWidgets(this)
            }
        }
        resetSplits()
        resetTimerPosition()
    }

    private fun resetSplits() {
        segmentTimes.clear()
        splitsIter = null
        mBinding.delta.visibility = View.GONE
        mBinding.currentSplit.visibility = View.GONE
    }

    private fun timerSplit() {
        splitsIter?.next()?.let {
            chronometer.compareAgainst = if (it.hasTime(comparison)) it.calculateSplitTime(comparison)
            else 0L
            mBinding.currentSplit.text = it.name
        }
    }

    private fun timerStart() {
        if (hasSplits) {
            splitsIter = category.splits.listIterator()
            timerSplit()
            mBinding.currentSplit.visibility =
                if (prefs.getBoolean(getString(R.string.key_pref_timer_show_current_split), true))
                    View.VISIBLE
                else View.GONE
        } else {
            chronometer.compareAgainst = category.bestTime
        }
        chronometer.start()
    }

    private fun timerStop() {
        chronometer.stop()
        mBinding.currentSplit.visibility = View.GONE
    }

    private fun getCurrentSplit(): Split? {
        splitsIter?.previous()
        return splitsIter?.next()
    }

    private fun updateDelta(currentSplit: Split, splitTime: Long) {
        val segmentTime = splitTime - currentSplitStartTime
        val delta = splitTime - currentSplit.calculateSplitTime(comparison)
        mBinding.delta.text = delta.getFormattedTime(plusSign = true)
        mBinding.delta.setTextColor(
            when {
                segmentTime < currentSplit.bestTime -> Chronometer.colorBestSegment
                delta < 0 -> Chronometer.colorAhead
                else -> Chronometer.colorBehind
            }
        )
        mBinding.delta.visibility = if (!currentSplit.hasTime(comparison)) View.GONE else View.VISIBLE
    }

    companion object {

        var IS_ACTIVE = false
        var gameName = ""
        var categoryName = ""

        private val scope = CoroutineScope(Dispatchers.Main)

        fun launchTimer(
            context: Context?,
            gameName: String,
            categoryName: String,
            minimizeIfNoGameLaunch: Boolean = true
        ) = scope.launch {
            context ?: return@launch
            if (IS_ACTIVE) {
                context.showToast(context.getString(R.string.toast_close_active_timer))
                return@launch
            }
            if (!context.tryLaunchGame(gameName)) {
                if (minimizeIfNoGameLaunch)
                    context.minimizeApp()
            }
            context.startTimerService(gameName, categoryName)
        }
    }
}
